test/jdk/java/foreign/nested/TestNested.java

```diff
@@ -0,0 +1,260 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @enablePreview
+ * @library ../ /test/lib
+ * @requires ((os.arch == "amd64" | os.arch == "x86_64") & sun.arch.data.model == "64") | os.arch == "aarch64" | os.arch == "riscv64"
+ * @requires vm.flavor != "zero"
+ * @build NativeTestHelper
+ * @run testng/othervm --enable-native-access=ALL-UNNAMED TestNested
+ */
+
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.GroupLayout;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SegmentAllocator;
+import java.lang.foreign.StructLayout;
+import java.lang.foreign.UnionLayout;
+import java.lang.invoke.MethodHandle;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class TestNested extends NativeTestHelper {
+
+    static {
+        System.loadLibrary("Nested");
+    }
+
+    @Test(dataProvider = "nestedLayouts")
+    public void testNested(GroupLayout layout) throws Throwable {
+        try (Arena arena = Arena.openConfined()) {
+            Random random = new Random(0);
+            TestValue testValue = genTestValue(random, layout, arena);
+
+            String funcName = "test_" + layout.name().orElseThrow();
+            FunctionDescriptor downcallDesc = FunctionDescriptor.of(layout, layout, C_POINTER);
+            FunctionDescriptor upcallDesc = FunctionDescriptor.of(layout, layout);
+
+            MethodHandle downcallHandle = downcallHandle(funcName, downcallDesc);
+            AtomicReference<Object[]> returnBox = new AtomicReference<>();
+            MemorySegment stub = makeArgSaverCB(upcallDesc, arena, returnBox, 0);
+
+            MemorySegment returned = (MemorySegment) downcallHandle.invokeExact(
+                    (SegmentAllocator) arena, (MemorySegment) testValue.value(), stub);
+
+            testValue.check().accept(returnBox.get()[0]);
+            testValue.check().accept(returned);
+        }
+    }
+
+    @DataProvider
+    public static Object[][] nestedLayouts() {
+        List<GroupLayout> layouts = List.of(
+                S1, U1, U17, S2, S3, S4, S5, S6, U2, S7, U3, U4, U5, U6, U7, S8, S9, U8, U9, U10, S10,
+                U11, S11, U12, S12, U13, U14, U15, S13, S14, U16, S15);
+        return layouts.stream().map(l -> new Object[]{l}).toArray(Object[][]::new);
+    }
+
+    static final StructLayout S1 = MemoryLayout.structLayout(
```

Comment:
Wouldn't it be better to have these structs defined inline inside the `List.of` call in the provider? That should make it a little easier to just add a new case.

Reply:
I don't think so, because there are struct and union layouts that are nested inside other layouts, but then _also_ are a test case themselves. So, I think we need some intermediate variables either way (or copy paste), and then it seems cleaner to do that for all the layouts.

Reply:
I see - missed that, thanks.